diff --git a/compiler/compiler.go b/compiler/compiler.go
index 6f8b89a..63d3325 100644
--- a/compiler/compiler.go
+++ b/compiler/compiler.go
@@ -1400,6 +1400,17 @@ func (b *builder) createFunctionStart(intrinsic bool) {
 		// It is necessary to pass a dummy alloca to runtime.trackPointer
 		// because runtime.trackPointer is replaced by an alloca store.
 		b.stackChainAlloca = b.CreateAlloca(b.ctx.Int8Type(), "stackalloc")
+
+		for _, param := range b.fn.Params {
+			if value, ok := b.locals[param]; ok && len(*param.Referrers()) != 0 {
+				b.trackValue(value)
+			}
+		}
+		for _, freeVar := range b.fn.FreeVars {
+			if value, ok := b.locals[freeVar]; ok && len(*freeVar.Referrers()) != 0 {
+				b.trackValue(value)
+			}
+		}
 	}
 }
 
diff --git a/transform/gc.go b/transform/gc.go
index 824aa23..c5d5eeb 100644
--- a/transform/gc.go
+++ b/transform/gc.go
@@ -165,12 +165,18 @@ func MakeGCStackSlots(mod llvm.Module) bool {
 			call.EraseFromParentAsInstruction()
 
 			// Some trivial optimizations.
-			if ptr.IsAInstruction().IsNil() {
+			isArgument := !ptr.IsAArgument().IsNil()
+			if ptr.IsAInstruction().IsNil() && !isArgument {
 				continue
 			}
 			if _, ok := rooted[ptr]; ok {
 				continue
 			}
+			if isArgument {
+				rooted[ptr] = struct{}{}
+				pointers = append(pointers, ptr)
+				continue
+			}
 			switch ptr.InstructionOpcode() {
 			case llvm.GetElementPtr:
 				// Check for all zero offsets.
@@ -287,7 +293,8 @@ func MakeGCStackSlots(mod llvm.Module) bool {
 		stackObjectType := ctx.StructType(fields, false)
 
 		// Create the stack object at the function entry.
-		builder.SetInsertPointBefore(fn.EntryBasicBlock().FirstInstruction())
+		entryFirst := fn.EntryBasicBlock().FirstInstruction()
+		builder.SetInsertPointBefore(entryFirst)
 		stackObject := builder.CreateAlloca(stackObjectType, "gc.stackobject")
 		initialStackObject := llvm.ConstNull(stackObjectType)
 		numSlots := (targetData.TypeAllocSize(stackObjectType) - uint64(targetData.PointerSize())*2) / uint64(targetData.ABITypeAlignment(uintptrType))
@@ -308,7 +315,10 @@ func MakeGCStackSlots(mod llvm.Module) bool {
 		pointerStores := make(map[llvm.Value]struct{})
 		for i, ptr := range pointers {
 			// Insert the store after the pointer value is created.
-			insertionPoint := llvm.NextInstruction(ptr)
+			insertionPoint := entryFirst
+			if ptr.IsAArgument().IsNil() {
+				insertionPoint = llvm.NextInstruction(ptr)
+			}
 			for !insertionPoint.IsAPHINode().IsNil() {
 				// PHI nodes are required to be at the start of the block.
 				// Insert after the last PHI node.
