diff --git a/src/internal/reflectlite/value.go b/src/internal/reflectlite/value.go
index 9a93e91..8d7aee0 100644
--- a/src/internal/reflectlite/value.go
+++ b/src/internal/reflectlite/value.go
@@ -2181,7 +2181,30 @@ func (v Value) FieldByIndex(index []int) Value {
 
 // FieldByIndexErr returns the nested field corresponding to index.
 func (v Value) FieldByIndexErr(index []int) (Value, error) {
-	return Value{}, &ValueError{Method: "FieldByIndexErr"}
+	if len(index) == 1 {
+		return v.Field(index[0]), nil
+	}
+	if v.Kind() != Struct {
+		return Value{}, &ValueError{"FieldByIndexErr", v.Kind()}
+	}
+	for i, x := range index {
+		if i > 0 {
+			if v.Kind() == Pointer && v.typecode.elem().Kind() == Struct {
+				if v.IsNil() {
+					return Value{}, nilEmbeddedError{}
+				}
+				v = v.Elem()
+			}
+		}
+		v = v.Field(x)
+	}
+	return v, nil
+}
+
+type nilEmbeddedError struct{}
+
+func (nilEmbeddedError) Error() string {
+	return "reflect: indirection through nil pointer to embedded struct field"
 }
 
 func (v Value) FieldByName(name string) Value {
diff --git a/src/os/exec.go b/src/os/exec.go
index 28406f9..eeb5f8b 100644
--- a/src/os/exec.go
+++ b/src/os/exec.go
@@ -3,6 +3,7 @@ package os
 import (
 	"errors"
 	"syscall"
+	"time"
 )
 
 var (
@@ -63,6 +64,10 @@ func (p *ProcessState) ExitCode() int {
 	return -1 // TODO
 }
 
+func (p *ProcessState) UserTime() time.Duration { return 0 }
+
+func (p *ProcessState) SystemTime() time.Duration { return 0 }
+
 type Process struct {
 	Pid int
 }
diff --git a/src/os/removeall_noat.go b/src/os/removeall_noat.go
index 4b37b3b..57a0ef6 100644
--- a/src/os/removeall_noat.go
+++ b/src/os/removeall_noat.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !baremetal && !js && !wasip1 && !wasip2 && !wasm_unknown && !nintendoswitch
+//go:build !baremetal && !js && !wasip2 && !wasm_unknown && !nintendoswitch
 
 package os
 
diff --git a/src/os/removeall_other.go b/src/os/removeall_other.go
index bf3265d..7feb915 100644
--- a/src/os/removeall_other.go
+++ b/src/os/removeall_other.go
@@ -1,4 +1,4 @@
-//go:build baremetal || js || wasip1 || wasip2 || wasm_unknown || nintendoswitch
+//go:build baremetal || js || wasip2 || wasm_unknown || nintendoswitch
 
 // Copyright 2009 The Go Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style
diff --git a/src/reflect/value.go b/src/reflect/value.go
index 6692669..9013de7 100644
--- a/src/reflect/value.go
+++ b/src/reflect/value.go
@@ -223,7 +223,7 @@ func (v Value) Method(i int) Value {
 }
 
 func (v Value) MethodByName(name string) Value {
-	panic("unimplemented: (reflect.Value).MethodByName()")
+	return Value{}
 }
 
 func (v Value) Recv() (x Value, ok bool) {
diff --git a/src/runtime/memprofile_stub.go b/src/runtime/memprofile_stub.go
new file mode 100644
index 0000000..c13acd8
--- /dev/null
+++ b/src/runtime/memprofile_stub.go
@@ -0,0 +1,3 @@
+package runtime
+
+var MemProfileRate int = 512 * 1024
diff --git a/src/runtime/runtime.go b/src/runtime/runtime.go
index 85684dd..a49ccd6 100644
--- a/src/runtime/runtime.go
+++ b/src/runtime/runtime.go
@@ -6,7 +6,7 @@ import (
 
 //go:generate go run ../../tools/gen-critical-atomics -out ./atomics_critical.go
 
-const Compiler = "tinygo"
+const Compiler = "gc"
 
 // Unit for the 'ticks' and 'sleepTicks' functions.
 //
